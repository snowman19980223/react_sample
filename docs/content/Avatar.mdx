---
title: Avatar
---


Avatars are images used to represent users and organizations on GitHub. They typically are squares with rounded edges.

## Default example

```jsx live
<Avatar mb={4} src="https://avatars.githubusercontent.com/primer" size={128} />

```
To create a Parent + Child avatar combination, wrap Avatars in AvatarPair:
```jsx live
 <AvatarPair my={4}>
  <Avatar src="https://avatars.githubusercontent.com/primer"/>
  <Avatar src="https://avatars.githubusercontent.com/primer"/>
 </AvatarPair>
 ```

## System props

Avatar components get `COMMON` system props. Read our [System Props](/system-props) doc page for a full list of available props.

## Component props

| Name | Type | Default | Description |
| :- | :- | :-: | :- |
| alt | String | | Passed through to alt on img tag |
| isChild | Boolean | | adds the `avatar-child` class if present |
| size | Number | 20 | adds the `avatar-small` class if less than 24 |
| src | String | | The source url of the avatar image |
