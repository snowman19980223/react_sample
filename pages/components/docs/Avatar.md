
# Avatar

Avatars are images used to represent users and organizations on GitHub. They typically are squares with rounded edges.

## Default example

```.jsx
<Avatar src="https://avatars.githubusercontent.com/primer" size={128} />
```

## System props

Avatar components get `COMMON` system props. Read our [System Props](/system-props) doc page for a full list of available props.

## Component props

| Name | Type | Default | Description |
| :- | :- | :-: | :- |
| alt | String | | Passed through to alt on img tag |
| isChild | Boolean | | adds the `avatar-child` class if present |
| size | Number | 20 | adds the `avatar-small` class if less than 24 |
| src | String | | The source url of the avatar image |

export const meta = {
  displayName: 'Avatar'
}
