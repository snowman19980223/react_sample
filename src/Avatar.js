import styled from 'styled-components'
import PropTypes from 'prop-types'
import sx from './sx'
import {get} from './constants'
import {space} from 'styled-system'
import systemPropTypes from '@styled-system/prop-types'
import theme from './theme'

function borderRadius({size}) {
  return {
    borderRadius: size <= 24 ? '2px' : '3px'
  }
}

const Avatar = styled.img.attrs(props => ({
  height: props.size,
  width: props.size,
  alt: props.alt
}))`
  display: inline-block;
  overflow: hidden; // Ensure page layout in Firefox should images fail to load
  line-height: ${get('lineHeights.condensedUltra')};
  vertical-align: middle;
  ${borderRadius};
  ${space};
  ${sx};
`

Avatar.defaultProps = {
  theme,
  size: 20,
  alt: ''
}

Avatar.propTypes = {
  alt: PropTypes.string.isRequired,
  size: PropTypes.number,
  src: PropTypes.string,
  ...systemPropTypes.space,
  ...sx.propTypes,
  theme: PropTypes.object
}

export default Avatar
