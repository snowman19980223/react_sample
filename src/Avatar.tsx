import PropTypes from 'prop-types'
import styled from 'styled-components'
import {COMMON, get, SystemCommonProps} from './constants'
import theme from './theme'
import sx, {SxProp} from './sx'

interface AvatarCustomProps {
  size: number,
  square: boolean,
  alt: string
}

function getBorderRadius(props: { size: number, square: boolean }) {
  if (props.square) {
    return props.size <= 24 ? '4px' : '6px'
  } else {
    return '50%'
  }
}

const Avatar = styled.img.attrs((props: AvatarCustomProps) => ({
  height: props.size,
  width: props.size,
  alt: props.alt
}))<AvatarCustomProps & SystemCommonProps & SxProp>`
  display: inline-block;
  overflow: hidden; // Ensure page layout in Firefox should images fail to load
  line-height: ${get('lineHeights.condensedUltra')};
  vertical-align: middle;
  border-radius: ${props => getBorderRadius(props)};
  ${COMMON};
  ${sx}
`

Avatar.defaultProps = {
  theme,
  size: 20,
  alt: '',
  square: false
}

Avatar.propTypes = {
  ...COMMON.propTypes,
  size: PropTypes.number,
  square: PropTypes.bool,
  ...sx.propTypes,
  theme: PropTypes.object
}

export type AvatarProps = React.ComponentProps<typeof Avatar>
export default Avatar
