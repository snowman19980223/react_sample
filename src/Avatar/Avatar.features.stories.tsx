import React from 'react'
import {ComponentMeta} from '@storybook/react'
import Avatar from './Avatar'

export default {
  title: 'Components/Avatar/Features',
  component: Avatar,
} as ComponentMeta<typeof Avatar>

export const Square = () => <Avatar square src="https://avatars.githubusercontent.com/primer" />

export const Size = () => (
  <div>
    <Avatar size={4} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
    <Avatar size={8} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
    <Avatar size={12} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
    <Avatar size={16} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
    <Avatar size={20} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
    <Avatar size={24} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
    <Avatar size={28} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
    <Avatar size={32} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
    <Avatar size={40} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
    <Avatar size={48} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
    <Avatar size={56} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
    <Avatar size={64} src="https://avatars.githubusercontent.com/u/92997159?v=4" />
  </div>
)
