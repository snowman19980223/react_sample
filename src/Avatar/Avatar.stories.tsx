import React from 'react'
import {ComponentMeta, Story} from '@storybook/react'
import Avatar, {AvatarProps, DEFAULT_AVATAR_SIZE} from './Avatar'
import {parseSizeFromArgs} from './storyHelpers'

export default {
  title: 'Components/Avatar',
  component: Avatar,
} as ComponentMeta<typeof Avatar>

type Args = {
  size?: number
  sizeAtNarrow?: number
  sizeAtRegular?: number
  sizeAtWide?: number
} & Omit<AvatarProps, 'size'>

export const Default = () => <Avatar src="https://avatars.githubusercontent.com/u/92997159?v=4" />

export const Playground: Story<Args> = args => {
  return (
    <Avatar
      size={parseSizeFromArgs(args)}
      square={args.square}
      src="https://avatars.githubusercontent.com/u/92997159?v=4"
    />
  )
}

Playground.args = {
  size: DEFAULT_AVATAR_SIZE,
}

Playground.argTypes = {
  size: {
    control: {
      type: 'number',
    },
  },
  sizeAtNarrow: {
    name: 'size.narrow',
    control: {
      type: 'number',
    },
  },
  sizeAtRegular: {
    name: 'size.regular',
    control: {
      type: 'number',
    },
  },
  sizeAtWide: {
    name: 'size.wide',
    control: {
      type: 'number',
    },
  },
  alt: {
    controls: false,
    table: {
      disable: true,
    },
  },
  ref: {
    controls: false,
    table: {
      disable: true,
    },
  },
  sx: {
    controls: false,
    table: {
      disable: true,
    },
  },
}
