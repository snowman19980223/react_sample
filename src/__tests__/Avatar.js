import React from 'react'
import Avatar from '../Avatar'
import renderer from 'react-test-renderer'

const render = component => renderer.create(component).toJSON()

describe('Avatar', () => {
  it('renders small by default', () => {
    expect(render(<Avatar />))
      .toEqual(render(<img className='avatar avatar-small' width={20} height={20} />))
  })

  it('respects the size prop', () => {
    expect(render(<Avatar size={40} />))
      .toEqual(render(<img className='avatar' width={40} height={40} />))
  })

  it('respects isChild', () => {
    expect(render(<Avatar isChild />))
      .toEqual(render(<img className='avatar avatar-small avatar-child' width={20} height={20} />))
  })

  it('passes through the src prop', () => {
    expect(render(<Avatar src='primer.png' />))
      .toEqual(render(<img className='avatar avatar-small' src='primer.png' width={20} height={20} />))
  })
})
