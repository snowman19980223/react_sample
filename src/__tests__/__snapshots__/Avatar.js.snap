// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders default props 1`] = `
<img
  alt=""
  blacklist={
    Array [
      "css",
      "m",
      "mt",
      "mr",
      "mb",
      "ml",
      "mx",
      "my",
      "p",
      "pt",
      "pr",
      "pb",
      "pl",
      "px",
      "py",
    ]
  }
  className="css-6rzz8m"
  height={20}
  width={20}
/>
`;
